### ISSUE TITLE
Load error on Fedora 29: Mutex creation failed

### ISSUE THREAD

**Issue by wch** [r-lib/later#77]
This is a followup to #63.

A Dockerfile that installs Anaconda and R on Fedora 29:

```R
FROM fedora:latest

RUN dnf install -y 'dnf-command(builddep)'
RUN dnf install -y wget bzip2

RUN wget https://repo.continuum.io/archive/Anaconda3-5.2.0-Linux-x86_64.sh
RUN bash ./Anaconda3-5.2.0-Linux-x86_64.sh -b
ENV PATH /root/anaconda3/bin:$PATH
RUN conda install r-base=3.5.1 r-essentials

CMD /bin/bash
```

To build the image and run the container, enter the directory with the Dockerfile and run:

```bash
docker build -t fr .
docker run --rm -ti fr /bin/bash
```

Inside the container, run:

```
R
options(repos=c(CRAN="https://cloud.r-project.org/"))
install.packages("later")
```

Some of the output:

```
> install.packages("later")
trying URL 'https://cloud.r-project.org/src/contrib/later_0.7.5.tar.gz'
Content type 'application/x-gzip' length 38900 bytes (37 KB)
==================================================
downloaded 37 KB

* installing *source* package ‘later’ ...
** package ‘later’ successfully unpacked and MD5 sums checked
Running configure script
Using CC=x86_64-conda_cos6-linux-gnu-cc
Using CFLAGS=-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -pipe -I/root/anaconda3/include -fdebug-prefix-map==/usr/local/src/conda/- -fdebug-prefix-map==/usr/local/src/conda-prefix
Using CPPFLAGS=-DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -I/root/anaconda3/include -Wl,-rpath-link,/root/anaconda3/lib
C11-style threads.h support not detected. Using tinycthread library.
...
...
...
x86_64-conda_cos6-linux-gnu-c++ -shared -L/root/anaconda3/lib/R/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,-rpath,/root/anaconda3/lib -L/root/anaconda3/lib -Wl,-rpath-link,/root/anaconda3/lib -o later.so RcppExports.o callback_registry.o init.o later.o later_native.o later_posix.o later_win32.o timer_posix.o timestamp_unix.o timestamp_win32.o -pthread ./tinycthread/tinycthread.o -L/root/anaconda3/lib/R/lib -lR
installing to /root/anaconda3/lib/R/library/later/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
sh: line 1:   387 Aborted                 (core dumped) '/root/anaconda3/lib/R/bin/R' --no-save --slave 2>&1 < '/tmp/RtmpkPZJXx/file10cd09078'
terminate called after throwing an instance of 'std::runtime_error'
  what():  Mutex creation failed
ERROR: loading failed
* removing ‘/root/anaconda3/lib/R/library/later’
* restoring previous ‘/root/anaconda3/lib/R/library/later’

The downloaded source packages are in
	‘/tmp/RtmpSEDqty/downloaded_packages’
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Warning message:
In install.packages("later") :
  installation of package ‘later’ had non-zero exit status
```

**Comment by wch** [comment]
Here's what I think is happening. In the configure phase, the later package tries to detect if the system supports C11-style threads. If it detects support for C11-style threads, then it simply uses the system's threads.h. If it does not detect support for C11-style threads, then it uses tinycthread.h, which mimics the API.

The problem is that Anaconda does not ship with threads.h, so the configure script will think that the system does not support C11-style threads, even if it does actually support them. This is the case with Fedora 29, where the system has glibc 2.28. During the linking phase, the calls to functions such as `mtx_init` (which are present in both tinycthread and C11-style threads) get linked to glibc's version of `mtx_init`, instead of the one the tinycthread version that later was compiled with.

In short: Anaconda does not supply threads.h, but glibc 2.28 has the functions defined in threads.h, so if any other code uses the same function names, they get linked to the glibc version of the function.

Even if the configure script properly detected that the system supported C11-style threads, Anaconda does not supply a threads.h to compile against.

cc @mingwandroid

**Comment by mingwandroid** [comment]
We do provide threads.h, it works.

Regardless of that, two autoloaded DSOs must never contain the same exported symbol names as otherwise which gets used is arbitrary.

So tinycthread needs to rename all the mtx_ functions if it's ever to work in the presence of a recent glibc where the program has already loaded pthreads.

**Comment by mingwandroid** [comment]
> Even if the configure script properly detected that the system supported C11-style threads

So my take on this (please correct me if I'm wrong) is it's not detecting C11 threads and tinycthread has become broken by recent changes to glibc.

**Comment by wch** [comment]
> So my take on this (please correct me if I'm wrong) is it's not detecting C11 threads and tinycthread has become broken by recent changes to glibc.

I think that's an accurate description.

@jcheng5 and I discussed what to do, and we think that what you suggested makes sense: rename the `mtx_` functions in tinycthread to, say, `tt_mtx_`, and use tinycthread unconditionally -- don't try to use threads.h at all. (I think the `cnd_` functions might also need to be renamed.) We can also:

* Get rid of the configure script.
* Replace `Makevars.in` with `Makevars`.
* Get rid of `c11threads.h` file, which basically just switches between threads.h and tinycthread.h.

@mingwandroid if you'd like to take a stab at the renaming, we'd appreciate it! You don't have to worry about the fiddly configure stuff if you don't want to -- we can take care of that.

**Comment by eddelbuettel** [comment]
Two more cents on removing `configure`:  In this [StackOverflow question](https://stackoverflow.com/questions/53558095/r-installation-of-package-later-fails) another Linux user (on an usual distro) sees things go sour; somehow `make` thinks `fopen` needs to be built (?).  By avoiding detection you end up being more resilient to what looks like faulty values being shipped. (he tagged that `rcpp` even though it has nothing to do with Rcpp but that's I got pulled-in.)

Otherwise, I just came here to suggest you add these two lines to allow debugging of `configure` straight in the shell via a fallback to get `R_HOME`.  I tested and if you go via the usual route it reflects alternate `RD` etc installations as it should.

```diff
edd@rob:/tmp/later$ diff -u later.orig/configure later/configure 
--- later.orig/configure        2018-08-29 17:17:53.000000000 -0500
+++ later/configure     2018-12-01 08:37:00.680322641 -0600
@@ -2,6 +2,8 @@
 
 echo "Running configure script"
 
+# Find R
+: ${R_HOME=`R RHOME`}
 # Find compiler 
 CC=`${R_HOME}/bin/R CMD config CC`
 CFLAGS=`${R_HOME}/bin/R CMD config CFLAGS`
edd@rob:/tmp/later$ 
```

**Comment by wch** [comment]
@mingwandroid Do you mind trying out PR #79?

@eddelbuettel In that PR I got rid of the configure script entirely. That should fix the issue you mentioned, right?

**Comment by mingwandroid** [comment]
Thanks @wch!

Tested on Anaconda Distribution with its compilers, macOS and Linux built fine and work as far as `library('later')` is concerned - can you recommend a more thurough test though? (I did not test Windows, shall I?)

**Comment by eddelbuettel** [comment]
@wch Quite possibly -- selection of R based on `${PATH}` should work.  The addition I suggestion basically brings that trick into `configure` (and was something Kurt once sent me for the same reason).

**Comment by mingwandroid** [comment]
@wch, would still like to get any suggestions you have for better testing of the Anaconda package if you have time.

**Comment by wch** [comment]
@mingwandroid You could run the unit tests for the package. To do this:

```
# Clone the repo
git clone https://github.com/r-lib/later.git

# Build source package
R CMD build --no-build-vignettes later

# Install with tests
R CMD INSTALL --install-tests later_0.7.5.9001.tar.gz

# Start R, run tests
install.packages('testthat')
library(testthat)
test_package('later')
```

It should give output like:

```
> test_package('later')
══ testthat results  ═══════════════════════════════════════════════════════════
OK: 17 SKIPPED: 0 FAILED: 0
```

**Comment by QuLogic** [comment]
Oh, hmm, that's unfortunate. I liked being able to just delete `tinycthreads.[ch]` to verify the Fedora package never used a bundled library. The renaming can be worked around, but can I get a promise that tinycthread won't be unnecessarily patched (except for critical issues, of course)?
